Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,404 @@
+        
+        guard !self.seenFirstResponse else {
+            // We're waiting for upgrade to complete: buffer this data.
+            self.receivedMessages.append(data)
+            return
+        }
+        
+        let responsePart = unwrapInboundIn(data)
+        
+        switch self.upgradeState {
+        case .awaitingConfirmationResponse:
+            self.firstResponseHeadReceived(context: context, responsePart: responsePart)
+        case .upgrading, .upgradingAddingHandlers:
+            if case .end = responsePart {
+                // This is the end of the first response. Swallow it, we're buffering the rest.
+                self.seenFirstResponse = true
+            }
+        case .upgraderReady(let upgrade):
+            if case .end = responsePart {
+                // This is the end of the first response, and we can upgrade. Time to kick it off.
+                self.seenFirstResponse = true
+                upgrade()
+            }
+        case .upgradeFailed:
+            // We were reentrantly called while delivering the response head. We can just pass this through.
+            context.fireChannelRead(data)
+        case .upgradeComplete:
+            //Upgrade has completed but we have not seen a whole response and still got reentrantly called.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.receivedResponseAfterUpgradeCompleted)
+        case .requestRequired:
+            //We are receiving an upgrade response and we have not requested the upgrade.
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.receivedResponseBeforeRequestSent)
+        }
+    }
+    
+    private func firstResponseHeadReceived(context: ChannelHandlerContext, responsePart: HTTPClientResponsePart) {
+        
+        // We should decide if we're can upgrade based on the first response header: if we aren't upgrading,
+        // by the time the body comes in we should be out of the pipeline. That means that if we don't think we're
+        // upgrading, the only thing we should see is a response head. Anything else in an error.
+        guard case .head(let response) = responsePart else {
+            context.fireErrorCaught(NIOHTTPClientUpgradeError.invalidHTTPOrdering)
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        // Assess whether the upgrade response has accepted our upgrade request.
+        guard case .switchingProtocols = response.status else {
+            self.notUpgrading(context: context, data: responsePart)
+            return
+        }
+        
+        // Ok, we have a HTTP response. Check if it's an upgrade confirmation.
+        // If it's not, we want to pass it on and remove ourselves from the channel pipeline.
+        let acceptedProtocols = response.headers[canonicalForm: "upgrade"].map(String.init)
+        
+        guard acceptedProtocols.count > 0 else {
```

Comment:
This check duplicates one that is in `handleUpgrade`: let's remove it in favour of that one.